docs/man/supybot-newplugin.1
.\" Process this file with
.\" groff -man -Tascii supybot-newplugin.1
.\"
.TH SUPYBOT\-NEWPLUGIN 1 "SEPTEMBER 2004"
.SH NAME
supybot\-newplugin \- A wizard for creating Supybot plugins
.SH SYNOPSIS
.B supybot\-newplugin
.RI [ options ]
.SH DESCRIPTION
.B
supybot\-newplugin
is a wizard that creates a template python source file for a new
.IR supybot (1)
plugin.
.SH OPTIONS
.TP
.B \-\^\-version
Show version of program.
.TP
.BR \-h ", " \-\^\-help
Show summary of options.
.TP
.BR \-r ", " \-\^\-regexp
Uses a regexp\-based callback.
.TP
.BI \-n " NAME" "\fR,\fP \-\^\-name=" NAME
Sets the name for the plugin.
.TP
.BR \-t ", " \-\^\-thread
Makes the plugin threaded.
.SH "SEE ALSO"
.IR python (1),
.IR supybot (1),
.IR supybot\-wizard (1),
.IR supybot\-adduser (1),
.SH AUTHOR
This manual page was originally written by James Vega
<vega dot james at gmail dot com>.  Permission is granted to copy,
distribute and/or modify this document under the terms of the Supybot
license, a BSD\-style license.


CHANGELOG
	* Implemented the FreeBSD module for searching ports.
	* Implemented the Unix module for *nix-specific stuff.
	* Implemented a few more commands for the Http module.
	* Changed names of modules to follow a standard convention.
	* Implemented @morehelp system.
	* Gave modules a .public attribute to determine if they should
	  be in @list.
	* Made threaded callbacks nestable :)
	* Removed 'commands' attribute/complication from irclib.Irc.

2002-Sep-03 Jeremy Fincher <jemfinch@users.sourceforge.net>

	* Version 0.60.0
	* Fixed TONS and TONS and TONS of bugs.
	* Included sample.conf, to show what a config file should look like.
	* Data for supybot/data has been separated out into its own package to
	decrease upload/download times.
	* A new Anagrams module has been added, as has a new Notes module.
	* All classes are new-style classes now.
	* I wrote a new reload function that both works more efficiently (no
	need to keep a cache of old objects anymore), more consistently (I can
	use new-style classes now), and in less code (it's much simpler the way
	it works).  This is a big gain.
	* I've written the beginnings of a ChannelStats class that handles the
	"seen" command, and later will collect information about different
	messages (how many JOINs, TOPICs, QUITs, etc.) and different stats like
	how many characters, words, messages sent to a channel in toto.
	* COMMAND NESTING R0X0RS!!!
	* Improved the telnet REPL slightly, a better login, printing, etc.
	* Improved the command-line version of the bot.
	* Added several new commands: xor, cfactive, version, crypt,
	timecommand, last, nslookup,

2002-Aug-27 Jeremy Fincher <jemfinch@users.sourceforge.net>

	* Version 0.5 -- We're at least halfway to 1.0 :)
	* Again, fixed TONS AND TONS AND TONS of bugs.
	* callbacks.Privmsg.reply is no longer implemented via an exception;
	subclasses can now self.reply more than once.
	* Minor API change in callbacks.Privmsg.getArgs; getArgsLastOptional
	has been removed in favor of the more-flexible optional keyword arg to
	getArgs.
	* Added several commands: exec, (un)setdefaultcapability, (un)settrace,
	calc (in FunCommands)
	* Added cdb library for constant databases that'll be independent of
	platform.
	* Added an asyncore-based interpreter via a socket that lets you telnet
	into the bot and execute code in the bot process.  Very useful for when
	I break things and need to reload but can't talk to the bot anymore :)
	* plugins.Forums is *really* useful :)
	* plugins.Http is getting useful :)
	* TONS of bugs fixed!
	* callbacks.Privmsg now has a "threaded" attribute that you can set if
	you want your subclass to spawn a new thread for every command.

2002-Aug-21 Jeremy Fincher <jemfinch@users.sourceforge.net>

	* Version 0.37.0 -- enough bugs to warrant a new minor release :)
	* Fixed ***TONS*** of bugs.
	* Added LICENSE to the top of all .py files written by me.
	* Added plugins/parter.py, which automatically parts a channel on JOIN.
	Useful for networks where, to determine who's a bot, they send everyone
	to a channel such as #botsmustgo and akill those who don't leave.
	* Added plugins/http.py, which hasn't been tested because I don't have
	threads on this Python install.
	* Made FunCommands.port work in reverse, too.  That's why the size of
	this tarball is significantly increased.
	* Added plugins/moobot.py, a duplication of some of the functionality
	of moobot, another bot written in Python.
	* Made ircutils.nick class, to support irc-case-insensitive comparison
	of nicks (and hashing) automatically.
	* Added unit testing framework for classes subclassing
	callbacks.Privmsg.

2002-Aug-07 Jeremy Fincher <jemfinch@user.sourceforge.net>

	* Version 0.36.3 (this header forgotten in the real release)
	* Ok, fine.  I'm using List Comprehensions now.  Happy?
	* Changed conf.makeClass to use imp instead of exec.
	* Changed traceback logging over to a modified text form of cgitb.
	* Added "find" to FactoidsCommands.
	* Added "capabilities" and "auth" and "unauth" to UserCommands.
	* Lots of little bugfixes.


2002-Jul-31 Jeremy Fincher <jemfinch@users.sourceforge.net>

	* Version 0.36.2
	* A whole lotta bug fixes, in general.
	* Fixed bug in FunCommands.timefun.
	* Added 'reconfig' command to OwnerCommands, to configuration changes
	can be enacted without restarting the bot.
	* Added 'protected' to the default turned-off capabilities in an
	IrcChannel.
	* Fixed callbacks.ChannelEnforcer so it wouldn't enforce MODES against
	the sender of the MODE itself).
	* Fixed username to make it lookup the hostmask of a nick given to it.
	* Fixed bug in sections command, and moved it to privmsgs.UserCommands
	since it's referenced by OwnerCommands.unload.
	* Fixed what I hope is the LAST bug in the help command.
	* Added users/channels default arguments to ircdb.checkIgnored and
	ircdb.checkCapability so they don't *have* to rely on global state.
	* Added CTCP SOURCE command.


2002-Jul-31 Jeremy Fincher <jemfinch@users.sourceforge.net>

	* Version 0.36.
	* Added makeFactoids.py and dumpFactoids.py to the tools/ subdirectory
	to facilitate the transfer of the bot from one Python installation to
	another.
	* Added Logger, a per-channel logger that logs rather similar to mIRC's
	channel format.
	* Fixed bug in RawLogger -- it was filtering away all messages from
	ever being sent to the network.
	* Added a factoids plugin.
	* Added __eq__ and __hash__ methods to ircmsgs.IrcMsg so it can be
	compared for equality and used in hash tables.  As a side note, I also
	changed .args to be a tuple rather than a list.
	* Added 'sections' command to FunCommands, to see what Privmsg-derived
	classes are hanging about in the bot.
	* Added 'unload' command to OwnerCommands, to remove a callback in a
	running Irc instance.
	* Added 'load' command to OwnerCommands, allowing owners to dynamically
	load a module that wasn't previously loaded and insert it into a
	running Irc instance.

scripts/supybot-newplugin.py
#!/usr/bin/env python

import supybot

import os

from questions import *

template = '''
#!/usr/bin/env python

###
# Copyright (c) 2002, Jeremiah Fincher
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#   * Redistributions of source code must retain the above copyright notice,
#     this list of conditions, and the following disclaimer.
#   * Redistributions in binary form must reproduce the above copyright notice,
#     this list of conditions, and the following disclaimer in the
#     documentation and/or other materials provided with the distribution.
#   * Neither the name of the author of this software nor the name of
#     contributors to this software may be used to endorse or promote products
#     derived from this software without specific prior written consent.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
###

"""
Add the module docstring here.  This will be used by the setup.py script.
"""

from baseplugin import *

import utils
import privmsgs
import callbacks


def configure(onStart, afterConnect, advanced):
    # This will be called by setup.py to configure this module.  onStart and
    # afterConnect are both lists.  Append to onStart the commands you would
    # like to be run when the bot is started; append to afterConnect the
    # commands you would like to be run when the bot has finished connecting.
    from questions import expect, anything, something, yn
    onStart.append('load %s')

example = utils.wrapLines("""
Add an example IRC session using this module here.
""")

class %s(%s):
    %s


Class = %s

# vim:set shiftwidth=4 tabstop=8 expandtab textwidth=78:
'''.strip() # This removes the newlines that precede and follow the text.

if __name__ == '__main__':
    name = something('What should the name of the plugin be?')
    if name.endswith('.py'):
        name = name[:-3]
    while name[0].islower():
        print 'Plugin names must begin with a capital.'
        name = something('What should the name of the plugin be?')
        if name.endswith('.py'):
            name = name[:-3]
    print textwrap.fill(textwrap.dedent("""
    Supybot offers two major types of plugins: command-based and
    regexp-based.  Command-based plugins are the kind of plugins
    you've seen most when you've used supybot.  They're also the
    most featureful and easiest to write.  Commands can be nested, 
    for instance, whereas regexp-based callbacks can't do nesting.
    
    That doesn't mean that you'll never want regexp-based callbacks.
    They offer a flexibility that command-based callbacks don't offer;
    however, they don't tie into the whole system as well.
    
    If you need to combine a command-based callback with some
    regexp-based methods, you can do so by subclassing
    callbacks.PrivmsgCommandAndRegexp and then adding a class-level
    attribute "regexps" that is a sets.Set of methods that are
    regexp-based.  But you'll have to do that yourself after this
    wizard is finished :)
    """).strip())
   
    if expect('Do you want a command-based plugin' \
              ' or a regexp-based plugin?',
              ['command', 'regexp']) == 'command':
        className = 'callbacks.Privmsg'
    else:
        className = 'callbacks.PrivmsgRegexp'
    print 'Sometimes you\'ll want a callback to be threaded.  If its methods'
    print '(command or regexp-based, either one) will take a signficant amount'
    print 'of time to run, you\'ll want to thread them so they don\'t block'
    print 'the entire bot.'
    print
    if yn('Does your plugin need to be threaded?') == 'y':
        threaded = 'threaded = True'
    else:
        threaded = 'pass'

    fd = file(name + '.py', 'w')
    fd.write(template % (name, name, className, threaded, name))
    fd.close()
    print 'Your new plugin template is in plugins/%s.py' % name

# vim:set shiftwidth=4 tabstop=8 expandtab textwidth=78:
